src/gc/gc.cpp

```diff
@@ -308,72 +308,55 @@ void GCStatistics::DisplayAndUpdate()
 
 #endif // GC_STATS
 
-#ifdef BIT64
-#define TOTAL_TIMES_TO_SHIFT 6
-#else
-#define TOTAL_TIMES_TO_SHIFT 5
-#endif // BIT64
-
 size_t round_up_power2 (size_t size)
 {
-    unsigned short shift = 1;
-    size_t shifted = 0;
-
-    size--;
-    for (unsigned short i = 0; i < TOTAL_TIMES_TO_SHIFT; i++)
-    {
-        shifted = size | (size >> shift);
-        if (shifted == size)
-        {
-            break;
-        }
-
-        size = shifted;
-        shift <<= 1;
-    }
-    shifted++;
-
-    return shifted;
+    // Get the 0-based index of the most-significant bit in the size.
+    // If the call failed (because `size` is zero), return 0.
+    DWORD msb_index;
+    if (0 ==
+#ifdef BIT64
+        BitScanReverse64(
+#else
+        BitScanReverse(
+#endif
+            &msb_index, size)) { return 0; }
+
+    // If `size` is already a power-of-two value, return it as-is.
+    // Otherwise, round it up to the next power-of-two by shifting
+    // it's MSB upwards one position.
+    size_t msb_mask = static_cast<size_t>(1) << msb_index;
+    // Always do this so the compiler uses e.g. cmov (x86) or csel (arm) below instead of a branch.
+    size_t rounded_up = msb_mask << 1;
+    return
+        msb_mask == size
+        ? size
+        : rounded_up;
 }
 
 inline
 size_t round_down_power2 (size_t size)
 {
     size_t power2 = round_up_power2 (size);
+    size_t rounded = power2 >> 1;
 
-    if (power2 != size)
-    {
-        power2 >>= 1;
-    }
-
-    return power2;
+    return power2 == size ? size : rounded;
```

Comment:
Why don't we directly call `BitScanReverse`, then return `msb_mask`?  For instance, for `size` equal to `0b1010` this code first calculates `msb_mask`, which equals to `0b1000`, then shifts it left one bit in `round_up_power2`, then finally shifts it right one bit in `round_down_power2`.  Returning `msb_mask` will avoid unnecessary shifts.

Reply:
I was leaning towards making this change as minimal as possible, but you're right -- it would be simpler and more efficient to just call ``BitScanReverse`` here.

Reply:
Thanks.  Note that now it works correctly even for values greater than `SIZE_MAX / 2` (which have `1` in the highest bit).